### ISSUE TITLE
Installation inconsistencies

### ISSUE THREAD

**Issue by rod-glover** [pacificclimate/crmprtd#107]
This issue deserves a better title; perhaps one will emerge after discussion and work on it.

I am trying and failing to install `crmprtd` on my workstation. I've encountered several kinds of difficulties, some of which are I think due to issues in the installation code. One may be unrelated to this project.

Here's what I've noted:

#### Pipfile error

[Pipfile erroneously adds specs for packages to be installed to the keys for the `pycds` package instead](https://github.com/pacificclimate/crmprtd/blob/master/Pipfile#L20-L27). This is easily enough fixed, and should be.

#### `Pipfile` and `setup.py` differ significantly

1. overall list of production packages 
   1. setup.py specifies only a subset of the packages named in Pipfile
   1. `Pipfile` specifies package `psycopg2-binary`; `setup.py` specifies `psycopg2`; see next section for more
1. overall list of dev/test packages (setup.py specifies `pysqlite`, which I think is not used, and omits others)

Installing `crmprtd` as a package executes `setup.py`, which the README advises to use for production environments. It also occurs in the Python CI installation, in conjunction with the pipenv-mediated install:

Python CI
```bash
pipenv install --deploy --dev
pipenv run pip install .  # [RG] why not `pipenv install .`?
```

#### `psycopg2-binary` vs `psycopg2`

1. Why is one specified in Pipenv (`psycopg2-binary`) and the other in setup.py (`psycopg2`)?
2. On my workstation, installing `psycopg2` fails with an error that suggests I need to install other libraries.

```
    In file included from psycopg/psycopgmodule.c:28:
    ./psycopg/psycopg.h:36:10: fatal error: libpq-fe.h: No such file or directory
       36 | #include <libpq-fe.h>
          |          ^~~~~~~~~~~~
    compilation terminated.
    
    It appears you are missing some prerequisite to build the package from source.

    You may install a binary package by installing 'psycopg2-binary' from PyPI.
```

I believe that any installation of `.` causes `setup.py` to be executed, which forces `psycopg2`, which fails, and for which a suggested remedy is 'psycopg2-binary'.

With `psycopg2` in `setup.py`, that alternative doesn't change things. 

If at all possible, I'd prefer to be able to install only `psycopg2-binary` and avoid `psycopg2`.

#### Prod installation 

```bash
pip install -i https://pypi.pacificclimate.org/simple crmprtd
# or with JSON logging functionality
pip install -i https://pypi.pacificclimate.org/simple crmprtd[jsonlogger]
```

Why isn't it something like `pipenv install crmprtd`?

#### Makefile vs Python CI installations

The Makefile installation and the Python CI installation differ in how they install the project, specifically in how they install `.`. Why?

Python CI
```bash
pipenv install --deploy --dev
pipenv run pip install .
```

Makefile (`install` target)
```bash
pipenv install --dev
pipenv install -e .  # [RG] does this run setup.py??
```

The Makefile is mentioned first in the README, presumably for dev installation.

**Comment by rod-glover** [comment]
There are 4 and maybe 5 separate but related questions/issues here. It may turn out that we should treat this as an "epic" and create separate GH issues for the different parts.

**Comment by jameshiebert** [comment]
You say "inconsistent"; I say "flexibility" :wink:

IMO there's nothing inherently wrong with having multiple ways to install software. But if you're having trouble find *a* right way, then I can understand your frustration.

> Pipfile error

Agreed. I found [another Pipfile error](https://github.com/pacificclimate/crmprtd/blob/master/Pipfile#L6) where there's a duplicate/wrong PyPI source. In previous versions of Pipenv, this was ignored, but now it errors out with an SSL error.

> Prod installation
> ...
> 
> Why isn't it something like pipenv install crmprtd?

Because pipenv is a PITA to install and use. And so for prod installation you don't have to have the entire source tree repository. The whole point of [building a package and uploading the source and binary distribution to a PyPI server](https://github.com/pacificclimate/crmprtd/blob/master/.github/workflows/pypi-publish.yml#L28) is so that it can be easily installed w/o all the code.

Note that when you install it, it grabs the pre-built wheel from our PyPI server:

```bash
hiebert@lynx:/local_temp/hiebert$ module load python/3.8.6
hiebert@lynx:/local_temp/hiebert$ python3.8 -m venv env
hiebert@lynx:/local_temp/hiebert$ source env/bin/activate
(env) hiebert@lynx:/local_temp/hiebert$ pip install -i https://pypi.pacificclimate.org/simple crmprtd[jsonlogger]
Looking in indexes: https://pypi.pacificclimate.org/simple
Collecting crmprtd[jsonlogger]
  Downloading https://pypi.pacificclimate.org/packages/crmprtd-3.5.1-py3-none-any.whl (266 kB)
     |████████████████████████████████| 266 kB 36.0 MB/s 
...
Successfully installed Mako-1.2.4 MarkupSafe-2.1.1 SQLAlchemy-1.4.44 alembic-1.8.1 bcrypt-4.0.1 black-22.10.0 certifi-2022.9.24 cffi-1.15.1 charset-normalizer-2.1.1 click-8.1.3 crmprtd-3.5.1 cryptography-38.0.3 geoalchemy2-0.12.5 greenlet-2.0.1 idna-3.4 importlib-metadata-5.0.0 importlib-resources-5.10.0 lxml-4.9.1 mypy-extensions-0.4.3 packaging-21.3 paramiko-2.12.0 pathspec-0.10.2 pint-0.20.1 platformdirs-2.5.4 psycopg2-2.9.5 pycds-3.3.0 pycparser-2.21 pynacl-1.5.0 pyparsing-3.0.9 pysftp-0.2.9 python-dateutil-2.8.2 python-json-logger-2.0.4 pytz-2022.6 pyyaml-6.0 requests-2.28.1 six-1.16.0 tomli-2.0.1 typing-extensions-4.4.0 urllib3-1.26.12 zipp-3.10.0
```

> psycopg2-binary vs psycopg2

> which fails, and for which a suggested remedy is 'psycopg2-binary'.

Can you just install the system package `libpq-dev` and move on with your life? These don't have to be mutually exclusive... one just needs to ensure that the target system can accommodate which version you choose. `libpq-dev` is an reverse dependency of `postgresql-server-dev-XX` so it's commonly installed on development systems...

```bash
hiebert@lynx:/local_temp/hiebert$ ls /usr/include/libpq*
/usr/include/libpq-events.h  /usr/include/libpq-fe.h

/usr/include/libpq:
libpq-fs.h
```

> Pipfile and setup.py differ significantly
> 
>     overall list of production packages
>         setup.py specifies only a subset of the packages named in Pipfile

I think that we've had this conversation repeatedly. `setup.py` lists *abstract* dependencies, or only packages that are *directly* used and imported by the package under development. A Pipfile specifies *concrete* dependencies, or the exact version of a package in an exact version of Python that is install-able. The former is by definition a subset of the latter, since concrete dependencies include subdependencies. IMO, there's nothing inherently wrong with these being "different".

> pipenv install --deploy --dev
> pipenv run pip install .  # [RG] why not `pipenv install .`?

Because `pipenv install` doesn't just install the named package into the environment. It also adds it to the Pipefile, and relocks and "updates" all of the dependencies. In short, it *changes the package under test*.

```bash
(env) hiebert@lynx:/local_temp/hiebert/crmprtd$ PIPENV_IGNORE_VIRTUALENVS=1 pipenv install .
Installing ....
Adding crmprtd to Pipfile's [packages]...
✔ Installation Succeeded 
Pipfile.lock (a773e2) out of date, updating to (707614)...
```

That's not what we want. We just want the package to be installed into the environment so that we can run the tests.

**Comment by rod-glover** [comment]
Thanks, James, this is very helpful. I take your point about flexibility. 

I have a few specific responses to some of your replies.

> > Prod installation
> > ...
> > Why isn't it something like pipenv install crmprtd?
> 
> Because pipenv is a PITA to install and use. And so for prod installation you don't have to have the entire source tree repository. The whole point of [building a package and uploading the source and binary distribution to a PyPI server](https://github.com/pacificclimate/crmprtd/blob/master/.github/workflows/pypi-publish.yml#L28) is so that it can be easily installed w/o all the code.

Yes, dumb question.

I do think that the two dev install methods should be grouped together, not separated by the prod method. It wasn't exactly clear what the Makefile method was for, although in retrospect it's fairly obvious it would be for dev install.

> > psycopg2-binary vs psycopg2
> 
> > which fails, and for which a suggested remedy is 'psycopg2-binary'.
> 
> Can you just install the system package `libpq-dev` and move on with your life? These don't have to be mutually exclusive... one just needs to ensure that the target system can accommodate which version you choose. `libpq-dev` is an reverse dependency of `postgresql-server-dev-XX` so it's commonly installed on development systems...

Point taken. It would also be helpful to have it named as prerequisite for an installation.

> > Pipfile and setup.py differ significantly
> > ```
> > overall list of production packages
> >     setup.py specifies only a subset of the packages named in Pipfile
> > ```
> 
> I think that we've had this conversation repeatedly. `setup.py` lists _abstract_ dependencies, or only packages that are _directly_ used and imported by the package under development. A Pipfile specifies _concrete_ dependencies, or the exact version of a package in an exact version of Python that is install-able. The former is by definition a subset of the latter, since concrete dependencies include subdependencies. IMO, there's nothing inherently wrong with these being "different".

Yes. The first comment re. subset was a lapse.

My second point, not addressed, but more important was:

> `Pipfile` specifies package `psycopg2-binary`; `setup.py` specifies `psycopg2`. 

_Any_ way of installing crmprtd, including via Pipenv, causes `setup.py` to be run, hence `psycopg2` to be installed. Perhaps `psycopg2-binary` has no utility and should be removed or replaced in the `Pipfile`?

> > pipenv install --deploy --dev
> > pipenv run pip install .  # [RG] why not `pipenv install .`?
> 
> Because `pipenv install` doesn't just install the named package into the environment. It also adds it to the Pipefile, and relocks and "updates" all of the dependencies. In short, it _changes the package under test_.
> 
> ```shell
> (env) hiebert@lynx:/local_temp/hiebert/crmprtd$ PIPENV_IGNORE_VIRTUALENVS=1 pipenv install .
> Installing ....
> Adding crmprtd to Pipfile's [packages]...
> ✔ Installation Succeeded 
> Pipfile.lock (a773e2) out of date, updating to (707614)...
> ```
> 
> That's not what we want. We just want the package to be installed into the environment so that we can run the tests.

Fair enough. The Python CI script _does_ use `pipenv install .`, and I often use it as a guide for how to install the package for testing. The difference raised questions in my mind. Possibly the CI script should be made consistent with the local install -- especially if it would result in difference in what packages are actually installed for testing, as you note above.

**Comment by jameshiebert** [comment]
> I do think that the two dev install methods should be grouped together, not separated by the prod method.

Yeah, that would make for clearer reading of the README, for sure.

> It wasn't exactly clear what the Makefile method was for, although in retrospect it's fairly obvious it would be for dev install.

Yeah... the Makefile is one of those things that is a nice idea in theory. In practice, I've always found it too rigid to be useful. Even in our own limited set of environments, it only works in Ubuntu/Debian and not Redhat/Fedora, which makes it so that we can't use it on *any* of our shared cloud systems.

> My second point, not addressed, but more important was:
> 
> > `Pipfile` specifies package `psycopg2-binary`; `setup.py` specifies `psycopg2`.
> 
> _Any_ way of installing crmprtd, including via Pipenv, causes `setup.py` to be run, hence `psycopg2` to be installed. Perhaps `psycopg2-binary` has no utility and should be removed or replaced in the `Pipfile`?

Yeah, I guess my point was that AFAIK `psycopg2` and `psycopg2-binary` are the same package, but with different installation media. One installs from source (and requires local dev headers) or one installs from a pre-built wheel. But you end up with the same code in your environment. If you want them to be consistent in our requirements, I can understand that. I don't care which one we use, but I'd say that the former provides greater flexibility at the expense of slightly inconvenient system dependencies.

> > > pipenv install --deploy --dev
> > > pipenv run pip install .  # [RG] why not `pipenv install .`?
> > 
> > 
> > Because `pipenv install` doesn't just install the named package into the environment. It also adds it to the Pipefile, and relocks and "updates" all of the dependencies. In short, it _changes the package under test_.
> > ```shell
> > (env) hiebert@lynx:/local_temp/hiebert/crmprtd$ PIPENV_IGNORE_VIRTUALENVS=1 pipenv install .
> > Installing ....
> > Adding crmprtd to Pipfile's [packages]...
> > ✔ Installation Succeeded 
> > Pipfile.lock (a773e2) out of date, updating to (707614)...
> > ```
> >   
> > That's not what we want. We just want the package to be installed into the environment so that we can run the tests.
> 
> Fair enough. The Python CI script _does_ use `pipenv install .`

I don't think that it does:

```bash
(env) hiebert@lynx:/local_temp/hiebert/crmprtd$ grep -r "pipenv install.*\." .github/workflows/*
(env) hiebert@lynx:/local_temp/hiebert/crmprtd$ 
```

That said, the Makefile and the README, both do:

```bash
(env) hiebert@lynx:/local_temp/hiebert/crmprtd$ grep -r "pipenv install.*\." *
Makefile:	pipenv install -e .
README.md:pipenv install --dev .
```

so maybe we can agree that they need to be updated to reflect actual practice? I also use the GH workflow files for guidance, since they are essentially an executable spec.

**Comment by rod-glover** [comment]
Oops, mixed up the Makefile/README with the CI script.

Apart from that, I think we are on much the same page about this now. Thanks for taking the time to help me straighten out my thinking, @jameshiebert .

**Comment by rod-glover** [comment]
Resolved in #109
